### PR TITLE
Finding Java source files

### PR DESCRIPTION
**Pull request by Avgor46**

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -102,6 +102,9 @@ Run casr-java:
 
     $ casr-java -o java.casrep -- java casr/tests/casr_tests/java/Test1.java
 
+You can specify path to Java source files via environment variable
+CASR\_SOURCE\_LOCATIONS  (list separated by ':').
```
Comment (anfedotoff) [inline docs/usage.md:106]: CASR_SOURCE_DIRECTORIES? We want directories not files.

File: libcasr/src/report.rs [hunk 2]
```diff
@@ -479,7 +479,22 @@ impl CrashReport {
             return None;
         }
 
-        if let Ok(file) = std::fs::File::open(&debug.file) {
+        // Search for file
+        let mut sources: Vec<PathBuf> = if let Ok(sources) = std::env::var("CASR_SOURCE_LOCATIONS")
```
Comment (anfedotoff) [inline libcasr/src/report.rs:483]: Directories, not files user should provide.

**Review by anfedotoff — CHANGES_REQUESTED** [review]

File: docs/usage.md (hunk 1 above)
Comment (Avgor46) [reply docs/usage.md]: done

File: libcasr/src/report.rs [hunk 3]
```diff
@@ -479,7 +479,30 @@ impl CrashReport {
             return None;
         }
 
-        if let Ok(file) = std::fs::File::open(&debug.file) {
+        // Search for file
```
Comment (anfedotoff) [inline libcasr/src/report.rs:482]: Move to casr-java. Just modify DebugInfo.file

File: docs/usage.md [hunk 4]
```diff
@@ -102,6 +102,9 @@ Run casr-java:
 
     $ casr-java -o java.casrep -- java casr/tests/casr_tests/java/Test1.java
 
+You can specify paths to directories with Java source files via environment variable
+CASR\_SOURCE\_DIRS  (list separated by ':').
```
Comment (SweetVishnya) [inline docs/usage.md:106]: We should set this variable in oss-sydr-fuzz/json-sanitizer Docker

Comment (SweetVishnya) [reply docs/usage.md:106]: Maybe, we can also use this variable for source files detection in sydr-fuzz cov-html?

File: casr/src/bin/casr-java.rs [hunk 5]
```diff
@@ -140,7 +141,30 @@ fn main() -> Result<()> {
...
+            };
+
+            if let Some(file) = source_dirs.iter().find_map(|dir| {
+                WalkDir::new(dir)
+                    .into_iter()
+                    .flatten()
+                    .map(|e| e.into_path())
+                    .filter(|e| e.is_file())
+                    .filter(|e| e.extension().is_some() && e.extension().unwrap() == "java")
+                    .find(|x| x.file_name().unwrap().to_str().unwrap().eq(&debug.file))
```
Comment (SweetVishnya) [inline casr/src/bin/casr-java.rs:163]: shouldn't we also get debug.file.file_name() to enable this search for arbitrary language (not only Java)?

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: casr/src/bin/casr-java.rs (hunk 5 above)
Comment (Avgor46) [reply casr/src/bin/casr-java.rs:163]: We use this search only for java. In java `debug.file` contains only filename.

File: docs/usage.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Avgor46) [reply docs/usage.md]: done

File: casr/src/bin/casr-java.rs (hunk 5 above)
Comment (SweetVishnya) [reply casr/src/bin/casr-java.rs:163]: debug.file is actually a full path to that file

Comment (SweetVishnya) [reply casr/src/bin/casr-java.rs:163]: It contains just a file name only in Java. However, we may write a more general case code.

Comment (SweetVishnya) [reply casr/src/bin/casr-java.rs:163]: debug.file -> debug.file.file_name()

File: libcasr/src/report.rs (hunk 3 above)
Comment (Avgor46) [reply libcasr/src/report.rs]: done

File: casr/src/bin/casr-java.rs (hunk 5 above)
Comment (Avgor46) [reply casr/src/bin/casr-java.rs]: done

**Review by anfedotoff — APPROVED** [review]

**Pull request merged**